### PR TITLE
prefer to InstanceExistsByProviderID rather than InstanceID

### PR DESCRIPTION
**Pull request by aoxn**
Signed-off-by: yaoyao.xyy <yaoyao.xyy@alibaba-inc.com>

**What this PR does / why we need it**:
Prefer to InstanceExistsByProviderID rather than InstanceID to keep consistency with cloudprovider interface specification.

### CODE REVIEW

File: pkg/controller/util/node/controller_utils.go [hunk 1]
```diff
@@ -171,20 +171,29 @@ func MarkAllPodsNotReady(kubeClient clientset.Interface, node *v1.Node) error {
...
-// ExistsInCloudProvider returns true if the node exists in the
-// cloud provider.
-func ExistsInCloudProvider(cloud cloudprovider.Interface, nodeName types.NodeName) (bool, error) {
+// EnsureNodeExistsByProviderID checks if the instance exists by the provider id,
+// If provider id in spec is empty it calls GenerateInstanceProviderID with node name to get provider id
+func EnsureNodeExistsByProviderID(cloud cloudprovider.Interface, node *v1.Node) (bool, error) {
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		var err error
+		providerID, err = cloudprovider.GenerateInstanceProviderID(context.TODO(), cloud, types.NodeName(node.Name))
```
Comment (andrewsykim) [inline pkg/controller/util/node/controller_utils.go]: `GenerateInstanceProviderID()` can never return `cloudprovider.InstanceNotFound`

Comment (aoxn) [reply pkg/controller/util/node/controller_utils.go]: Yes, removed cloudprovider.InstanceNotFound assertion.

**Comment by aoxn** [comment]
@ingvagabund @deads2k Do you have any comment on this, pls?

File: pkg/controller/util/node/controller_utils.go [hunk 2]
```diff
@@ -171,20 +171,30 @@ func MarkAllPodsNotReady(kubeClient clientset.Interface, node *v1.Node) error {
...
-		return false, fmt.Errorf("%v", ErrCloudInstance)
+		return false, errors.New("Instance interface Not implemented")
 	}
-	if _, err := instances.InstanceID(context.TODO(), nodeName); err != nil {
-		if err == cloudprovider.InstanceNotFound {
-			return false, nil
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		var err error
+		providerID, err = instances.InstanceID(context.TODO(), types.NodeName(node.Name))
```
Comment (andrewsykim) [inline pkg/controller/util/node/controller_utils.go]: `instanceID, err`

**Comment by andrewsykim** [comment]
@cheftako can we have your review here? It consolidates "does node exist" logic between KCM and CCM. We probably want to hold on this PR until https://github.com/kubernetes/kubernetes/pull/68310 merges but will leave it up to you

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 3]
```diff
@@ -725,7 +724,7 @@ func (nc *Controller) monitorNodeStatus() error {
 					}
 					continue
 				}
-				exists, err := nc.nodeExistsInCloudProvider(types.NodeName(node.Name))
+				exists, err := nc.nodeExistsInCloudProvider(node)
```
Comment (cheftako) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: We should work with sig-node to remove this case from this controller. It is the only piece of cloud specific code in this controller. If we remove this call then the controller becomes CP agnostic.

**Review by cheftako — COMMENTED** [review]
/lgtm

**Pull request closed without merging**